src/core/utils/flash.cpp

```diff
@@ -225,71 +522,97 @@ bool Flash::DoesValidRecordExist(uint32_t aOffset, uint16_t aKey) const
             continue;
         }
 
-        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, record.GetSize());
-        otPlatFlashWrite(&GetInstance(), dstIndex, dstOffset, &record, record.GetSize());
+        record.LoadData(&GetInstance(), mSwapIndex, srcOffset);
+#if OPENTHREAD_CONFIG_FLASH_LEGACY_COMPATIBILITY
+        record.Fixup();
+#endif
+        record.Save(&GetInstance(), dstIndex, dstOffset);
         dstOffset += record.GetSize();
     }
 
 exit:
-    otPlatFlashWrite(&GetInstance(), dstIndex, 0, &sSwapActive, sizeof(sSwapActive));
-    otPlatFlashWrite(&GetInstance(), mSwapIndex, 0, &sSwapInactive, sizeof(sSwapInactive));
+    swapHeader.Init(eraseCounter);
+    swapHeader.Save(&GetInstance(), dstIndex);
 
-    mSwapIndex = dstIndex;
-    mSwapUsed  = dstOffset;
+    swapHeader.Load(&GetInstance(), mSwapIndex);
+#if OPENTHREAD_CONFIG_FLASH_LEGACY_COMPATIBILITY
+    if (swapHeader.GetFormat() == kFormatV1)
+    {
+        otPlatFlashErase(&GetInstance(), mSwapIndex);
+    }
+    else
+#endif
+    {
+        swapHeader.SetInactive();
+        swapHeader.Save(&GetInstance(), mSwapIndex);
+    }
+
+    mSwapIndex      = dstIndex;
+    mSwapUsed       = dstOffset;
+    mSwapHeaderSize = sizeof(swapHeader);
 }
 
 otError Flash::Delete(uint16_t aKey, int aIndex)
 {
     otError      error = OT_ERROR_NOT_FOUND;
-    int          index = 0; // This must be initalized to 0. See [Note] below.
-    RecordHeader record;
+    int          index = 0;
+    RecordHeader recordHeader;
+    uint16_t     size;
 
-    for (uint32_t offset = kSwapMarkerSize; offset < mSwapUsed; offset += record.GetSize())
+    for (uint32_t offset = mSwapHeaderSize; offset < mSwapUsed; offset += size)
     {
-        otPlatFlashRead(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+        recordHeader.Load(&GetInstance(), mSwapIndex, offset);
+
+        size = recordHeader.GetSize();
 
-        if ((record.GetKey() != aKey) || !record.IsValid())
+        if (recordHeader.GetKey() != aKey)
         {
             continue;
         }
 
-        if (record.IsFirst())
+        if (recordHeader.InvalidatesPrevious())
         {
             index = 0;
+            error = OT_ERROR_NOT_FOUND;
         }
 
-        if ((aIndex == index) || (aIndex == -1))
+        if (!recordHeader.IsValid())
         {
-            record.SetDeleted();
-            otPlatFlashWrite(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
-            error = OT_ERROR_NONE;
+            continue;
         }
 
-        /* [Note] If the operation gets interrupted here and aIndex is 0, the next record (index == 1) will never get
-         * marked as first. However, this is not actually an issue because all the methods that iterate over the
-         * settings area initialize the index to 0, without expecting any record to be effectively marked as first. */
-
-        if ((index == 1) && (aIndex == 0))
+        if ((aIndex == index) || (aIndex == -1))
```

Comment:
In the case of `aIndex == -1`, I think it would be better to atomically delete all existing records. I think the best way to do this is by append a record that indicates all previous are invalidated. Thoughts?